### PR TITLE
Improving Filter Box

### PR DESCRIPTION
**Pull request by mistercrunch**
There's a longer term plan to improve filters, but this PR is aiming at adding a few well-needed options to the old "Filter Box". I figured the `FilterCollectionControl` can be reused in the future.

related to https://github.com/apache/incubator-superset/issues/5925, https://github.com/apache/incubator-superset/issues/5269, https://github.com/apache/incubator-superset/issues/3272, https://github.com/apache/incubator-superset/issues/2367

screen shot 2018-12-12 at 8 40 50 am

TODO:
* unit test for `FilterBoxItemControl`
* unit test for `FormRow`
* parallelize server side?

### CODE REVIEW

File: superset/assets/src/chart/ChartRenderer.jsx [hunk 1]
```diff
@@ -38,7 +38,7 @@ const defaultProps = {
   triggerRender: false,
 };
 
-class ChartRenderer extends React.PureComponent {
+class ChartRenderer extends React.Component {
```
Comment (mistercrunch) [inline superset/assets/src/chart/ChartRenderer.jsx]: Unrelated change, caught some React related warning

**Comment by vylc** [comment]
Tweaked the verbiage for some of these options
screen shot 2018-12-21 at 2 46 43 pm

**Comment by mistercrunch** [comment]
Another idea I had was to support multi-columns eventually. That'd be how we'd support "cascading dropdowns" , powered by a single query.

**Comment by mistercrunch** [comment]
What's the `Global` option?

**Comment by mistercrunch** [comment]
screen shot 2018-12-28 at 10 09 07 am

**Comment by mistercrunch** [comment]
Hey I think I'd like to checkpoint the work on filters here. We can improve things further in a future PR but this is already a fairly large PR, and closes many features request as is.

**Comment by kaitorecca** [comment]
How it handle to make filter impact on other chart component on the same default when default value is set? Per my understanding from code,  once you set value, it just show there, not effect other chart because only onchange event was there

**Comment by mistercrunch** [comment]
good catch @kaitorecca, opened an issue here: https://github.com/apache/incubator-superset/issues/6746

File: superset/migrations/versions/fb13d49b72f9_better_filters.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    filter_box_slices = session.query(Slice).filter_by(viz_type='filter_box')
+    for slc in filter_box_slices.all():
+        try:
+            params = json.loads(slc.params)
+            logging.info(f'Upgrading {slc.slice_name}')
+            cols = params.get('groupby')
+            metrics = params.get('metrics')
```
Comment (michellethomas) [inline superset/migrations/versions/fb13d49b72f9_better_filters.py]: @mistercrunch it looks like the param for filter_box was `metric` so this isn't migrating any of the filter box metrics correctly, they are getting set to `null` cc: @graceguo-supercat

**Comment by michellethomas** [comment]
@mistercrunch another issue, the new filters break with filter charts on druid datasources if you don't select a metric in your filter. Metric is now not a required field, and it seems sqla datasources don't require you to have a metric, but filter boxes on druid datasource will break without a metric with `list index out of range`
```
File "superset/connectors/druid/models.py", line 1150, in run_query
    order_by = list(qry['aggregations'].keys())[0]
IndexError: list index out of range
```

Can we also add tests for this case? 
cc @graceguo-supercat

**Comment by mistercrunch** [comment]
👀

File: superset/migrations/versions/fb13d49b72f9_better_filters.py (hunk 2 above)
Comment (mistercrunch) [reply superset/migrations/versions/fb13d49b72f9_better_filters.py]: You're right, and I'm confused as to how this got through... Let me send a fix now

**Comment by mistercrunch** [comment]
@michellethomas thanks for reporting this, issue, I submitted a fix here https://github.com/apache/incubator-superset/pull/7066
